yourbench/pipeline/single_shot_question_generation.py

```diff
@@ -0,0 +1,489 @@
+    - The generated questions are intended to be standalone, moderately challenging,
+      and reflect a deep understanding of the provided text chunk.
+
+Usage:
+    1) The pipeline will call the `run()` function from this module if the user configures
+       `pipeline.single_shot_question_generation.run = True`.
+    2) This function loads the required dataset (specified in the pipeline configuration),
+       samples chunks if necessary, and calls an LLM to generate questions.
+    3) The output is stored in a new dataset containing each generated question,
+       an estimated difficulty rating, and the model's self-provided reasoning.
+
+Stage-Specific Logging:
+    - All errors and relevant log messages are recorded in `logs/single_shot_question_generation.log`.
+
+Google-Style Docstrings:
+    - This codebase uses Python type hints and Google-style docstrings for clarity,
+      maintainability, and consistency.
+"""
+
+import json
+import os
+import random
+import re
+from dataclasses import dataclass
+from typing import Any, Dict, List
+
+from datasets import Dataset
+from loguru import logger
+
+from yourbench.utils.dataset_engine import (
+    save_dataset,
+    smart_get_output_dataset_name,
+    smart_get_output_subset,
+    smart_get_source_dataset_name,
+    smart_get_source_subset,
+    smart_load_dataset,
+)
+from yourbench.utils.inference_engine import InferenceCall, run_inference
+from yourbench.utils.prompts import (
+    QUESTION_GENERATION_SYSTEM_PROMPT,
+    QUESTION_GENERATION_USER_PROMPT,
+)
+
+
+# Ensure the logs directory exists and set up a stage-specific log file.
+os.makedirs("logs", exist_ok=True)
+logger.add("logs/single_shot_question_generation.log", level="DEBUG")
+
+
+@dataclass
+class SingleHopQuestionRow:
+    """
+    Represents a single-hop question row derived from a single chunk of text.
+
+    Attributes:
+        chunk_id: A string identifier for the chunk from which this question was generated.
+        document_id: Identifier for the parent document.
+        question: The generated question text.
+        self_answer: The LLM-produced short answer or reasoning.
+        estimated_difficulty: An integer from 1-10 indicating the estimated difficulty.
+        self_assessed_question_type: A descriptor for the type or style of question.
+        generating_model: The model used to generate this question.
+        thought_process: Free-form text describing how the question was derived or the
+            model's chain-of-thought (if provided).
+        raw_response: The full, unedited response from the model.
+    """
+
+    chunk_id: str
+    document_id: str
+    question: str
+    self_answer: str
+    estimated_difficulty: int
+    self_assessed_question_type: str
+    generating_model: str
+    thought_process: str
+    raw_response: str
+    citations: List[str]
+
+
+def run(config: Dict[str, Any]) -> None:
+    """
```

Comment:
Same comment as for the other - think about your "run" functions as needing to be very legible, by using unit subfunctions for each feature